Add Bessel function of the first kind of order one, besselj1

rreusser opened the pull request:
Resolves #29 .

<!--lint disable first-heading-level-->

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

- Adds the bessel function of the first kind of order one.

## Related Issues

> Does this pull request have any related issues?

No.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

<!-- <links> -->

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

<!-- </links> -->

rreusser committed:
Add Bessel function of the first kind of order one, besselj1

rreusser committed:
Tighten thresholds on besselj1 tests

rreusser committed:
Fix bessel function id in link database

rreusser commented:
Note: `bessel-j0` turned out to be a poor choice of an id for bessel functions of the first kind. They're not all zeroth order. `bessel-j` seems better.

rreusser commented:
Also, question: which is better for a readme title: `j1`, `besselj1` or something prettier?

rreusser committed:
Fix cpp-style comment in besselj1

Planeshifter commented:
Julia uses the `bessel` prefix, while Python does not. Both seem viable choices to me, although I prefer the longer version as it is more descriptive. Your thoughts, @kgryte?

Planeshifter commented:
Addendum: Since the question was solely about the README title and we have already settled on `j0` as the variable name for the function, sticking with `j0` seems fine to me.

kgryte commented:
`j1` is fine for the README title. The pkg name is fine as `besselj1`.

kgryte commented on docs/links/database.json:
```diff
@@ -1342,7 +1342,7 @@
   },
   "https://en.wikipedia.org/wiki/Bessel_function#Bessel_functions_of_the_first_kind:_J.CE.B1": {
-    "id": "bessel-j0",
+    "id": "bessel-j",
```
Alternatively, this could be more descriptive: `bessel-first-kind`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/README.md:
```diff
@@ -0,0 +1,71 @@
+
+#### j1( x )
+
+Compute the [Bessel function of the first kind][bessel-j] of order one at `x`.
```
s/Compute/Computes/

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/benchmark/cpp/benchmark.boost.cpp:
```diff
@@ -0,0 +1,116 @@
+	mt19937 rng;
+
+	// Define a uniform distribution for generating pseudorandom numbers as "doubles" between a minimum value (inclusive) and a maximum value (exclusive):
+	uniform_real_distribution<> randu( -100.0, 100.0 );
```
The value regime should match that of `C` and other benchmarks.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/benchmark/r/benchmark.R:
```diff
@@ -0,0 +1,93 @@
+	#' elapsed <- benchmark( 10000L );
+	benchmark <- function( iterations ) {
+		# Run the benchmarks:
+		results <- microbenchmark::microbenchmark( besselJ( runif(1)*100.0, 1 ), times = iterations );
```
Value regime should match that of `C` and other benchmarks. The reason, conceptually, is to have the primary "independent variable" be language and prevent some benchmarks from overly representing certain value ranges over others which may put that language at an inherent disadvantage. By keeping test values the same, we hopefully treat all langs more or less the same.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/lib/index.js:
```diff
@@ -0,0 +1,36 @@
+
+// MODULES //
+
+var j1 = require('./j1.js');
```
Spaces. I already addressed this in `j0`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/lib/j1.js:
````diff
@@ -0,0 +1,213 @@
+		* value = f * ( rc * cos( z ) - y * rs * sin( z ) );
+		* ```
+		*
+		* but using the sin/cos addition rules plus constants
````
Remove line wrapping.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 600.0;
```
Can you put the scale factor first? e.g., `600.0 * EPS`. Doing so makes it easier to scan by eye the tolerance magnitude and also read `600eps` LTR, rather than RTL with noise in between. I have already addressed this in `j0`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 300.0;
```
Scale factor first.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 2600.0;
```
Scale factor first.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 5.0;
```
Scale factor first.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 5.0;
```
Scale factor first.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 3.0;
```
Scale factor first.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,270 @@
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = EPS * abs( expected[i] ) * 3000.0;
```
Scale factor first.

kgryte commented:
There was a doc bug in `develop` introduced by me when renaming a link identifier. If you could merge in the latest develop, that would be great. :)

kgryte commented:
Can you also add an equation to the README? Just use the equation comment snippet which can be found in the snippets directory in tools. No need to add the inner `div` or actual equation SVG, as these will be generated from the equation comment during a separate build process. The same goes for `bessely0`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/README.md:
````diff
@@ -0,0 +1,71 @@
+
+## Usage
+
+``` javascript
````
Heads up: due to how a dependency (`remark`) generates Markdown, our Markdown style has updated a bit. Notably, a space no longer exists between the fence and the language. To avoid a needless diff in future builds, you can remove the space ahead-of-time and spare file churn.

cc @Planeshifter

rreusser commented:
This PR is for `besselj1`, but to get large numbers correct, I was looking at the asymptotic form from A&S, which applies to all bessel function orders. So working through `j0` at `x = 1e10`:

```
wolfram:        2.1755917502468917268590552836382091970671228212154530 × 10^-6
js besselj0:    2.175591750246892e-6
js asymptotic:  2.175591750246892e-6
julia:          2.1755917502468914e-6
scipy.j0:       2.1755892946936445e-06
boost:          2.175591750345720e-06
```

So not too bad. At `1e20`:

```
wolfram:        6.6980090407034242848762363480003437539735122779314444 × 10^-12
js besselj0:    6.698009040703425e-12
js asymptotic:  6.698009040703419e-12
julia:          6.698009040703426e-12
scipy.j0:       6.0956019061437597e-11
boost:         -7.9654185661602431e-11
```

Whoopsie. At `x = 1e30`:

```
wolfram:       -6.127370231742228805683191859553352335386736007371264 × 10^-16
js besselj0:   -5.589003016686146e-16
js asymptotic: -5.589003016686146e-16
julia:         -5.589003016686147e-16
scipy.j0:      -7.9784982162396589e-16
boost:         -2.0869531125378018e-17
```

And finally at `1e308`:

```
wolfram: J_0(100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
js besselj0:   -2.470656412079008e-155
js asymptotic: -2.4706564120790077e-155
julia:         -2.470656412079008e-155
scipy.j0:      -7.1116164028608716e-155
boost:          5.7780252703433500e-155
```

LOL wolfram.

So I guess kinda where I'm going with this is that by the time you get to 1e30 or so, the magnitude seems mostly fine but the phase is probably utterly meaningless. To the point that you probably get what you deserve if you evaluate oscillatory functions there. The bessel functions obviously *have* a single well-defined value at these magnitudes, but I'd lose no sleep shipping code that returns values with these sorts of accuracies. Agreement with julia is good at least.

rreusser commented:
Sorry, so to summarize: I think the plain no-asymptotic-approximation approach is fine for `besselj0` and `besselj1`, even at *very* large magnitudes. I thought maybe the non-asymptotic form would be necessary for `besselj1`, but it was actually worse and resulted in overflow until I fixed the associativity of the floating point calculations. With that fixed, they're both fine and non-asymptotic seems lovely.

For `besselj`, normal everyday magnitudes will take some work implementing a fancy approach. That part is not done. Small magnitudes will require an asymptotic form at the origin. Large magnitudes can use the asymptotic approximation I've been testing, which I think is perfectly fine, accuracy-wise.

Oh, and regarding overflow issues, the asymptotic version has the general form:

```
x = 1e308
q = x * x * x * x * x * ...                    ( = Infinity )
phase = multplier * ( 1.0 + numerator / q )    ( = multiplier )
```

So in short, overflow doesn't turn out to be an issue (again, once I fixed silly associativity issues on my end). It overflows like crazy, but it's all in the denominator and so just resolves itself.

kgryte commented:
@rreusser Awesome!!! Thanks for all this! Super informative.

__Aside__: what is with the `boost` values being all over the place?

rreusser commented:
@kgryte: re: boost, not sure. I was instrumenting the source to figure out which code paths it was executing to get the answers. I don't think I broke anything. I've been copying that code basically directly, so I was a bit surprised. I'm not sure. Could be worth investigating, though the agreement with julia is adequately satisfying for me overall.

kgryte commented:
@rreusser Agreed. Seems fine to me at this point.

kgryte commented:
@rreusser At this point, what is left to do for `besselj0`, `bessely0`, and `besselj1`?

rreusser commented:
- [x] add tests covering full gamut of numbers for `besselj0`
- [x] add tests covering full gamut of numbers for `besselj1`
- [x] fix `besselj1` PR review issues
- [x] implement large x asymptotic form for `besselj` (see: [bessel_jy_asym.hpp](https://github.com/boostorg/math/blob/242e9d669198415d165bb0213f730c00985bee9a/include/boost/math/special_functions/detail/bessel_jy_asym.hpp))
- [ ] implement small x asymptotic form for `besselj` (see: [bessel_jy_series.hpp](https://github.com/boostorg/math/blob/242e9d669198415d165bb0213f730c00985bee9a/include/boost/math/special_functions/detail/bessel_jy_series.hpp))
- [ ] implement medium-x form for computing general `besselj(v, x)` (which computes `besselj` and `bessely` at the same time: see: [bessel_jy.hpp](https://github.com/boostorg/math/blob/develop/include/boost/math/special_functions/detail/bessel_jy.hpp))
- [ ] delegate integer-valued `besselj(n, x)` to `besselj0` and `besselj1`
- [ ] test `besselj`
- [ ] implement `bessely0` (easy? analogous to `besselj0`?)
- [ ] implement `bessely1` (easy? analogous to `besselj1`?)
- [ ] implement `besselyn` (difficult part addressed already in `bessel_jy`?)

kgryte commented:
Thanks for the rundown.

rreusser committed:
Fix issues in besselj1

rreusser committed:
Rename link id bessel-j to bessel-first-kind

rreusser committed:
Fix R benchmark range in besselj1

rreusser commented:
An additional note so that it's documented somewhere:

Accuracy for subnormal x is very poor. Accuracy seems fine at `x = 1e-308` and goes progressively to zero at `5e-324`. This suggests to me that overflow (a reciprocal involved?) is steadily chopping off digits until the computation is worthless at `5e-324`. I think this could be overcome but given the strong disagreement at much less exotic magnitudes, I don't think it's of particular value to resolve this now.

kgryte commented:
Would also be worth placing a note in the source code in an extended JSDoc comment (e.g., in the function signature comment for the main export).

rreusser committed:
Merge branch 'develop' into besselj1

rreusser commented:
Main export = `lib/index.js` or `lib/j1.js`?

kgryte commented:
What are you referring to? In `package.json` or filename or something else?

rreusser commented:
Sorry, to your comment on adding a note in the source code "(e.g., in the function signature comment for the main export)". Specifically having this note in`j1.js` seems more relevant, in `index.js` seems more visible.

kgryte commented:
Ah, right. In `j1.js`, which is where the function signature and method and other details are usually added in extended comments. In `index.js`, we typically only articulate the module definition.

rreusser committed:
Add comment on accuracy of besselj1 for subnormal input

rreusser commented:
Thanks, added.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,92 @@
+
+# The full gamut of negative magnitudes:
+x = [ -( 10.0 ^ x ) for x in linspace( -308.0, 308.0, 317 ) ];
+print( x )
```
Remove this print statement.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/docs/repl.txt:
```diff
@@ -0,0 +1,30 @@
+    Returns
+    -------
+    y: number
+        Bessel function of the first kind of order one.
```
What is returned is not so much the "Bessel function" but the function value. I would just replace with `Function value.`. (Aside: I have already updated `besselj0` similarly.)

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/lib/j1.js:
```diff
@@ -0,0 +1,215 @@
+*
+* ## Notes
+*
+* -   Accuracy for subnormal x is very poor. Full accuracy is achieved at 1.0e-308 but trends progressively to zero at 5e-324. This suggests that underflow (or overflow, perhaps due to a reciprocal) is effectively cutting off digits of precision until the computation is worthless at 5e-324.
```
s/x/`x`/

In comments like these, we typically markup variables, etc, with backticks.

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/test/test.js:
```diff
@@ -0,0 +1,319 @@
+	t.end();
+});
+
+tape( 'the function evaluates the Bessel function of the first kind of order one (J_1) of `x` on the interval `[50,500]`', function test( t ) {
```
Rather than specify the exact intervals, describe in prose. The intervals may change but the intent will not.

> the function evaluates the Bessel function of the first kind of order one (J_1) (very large positive values)

Aside: I have updated `besselj0` similarly.

kgryte replied:
The same applies for other test descriptions.

rreusser replied:
Thanks. I've done this for the remaining bessel modules in the works.

kgryte replied:
Awesome. Thanks!

rreusser committed:
Fix review issues for besselj1

rreusser committed:
Fix alt tag in besselj1 equation

rreusser committed:
Rename besselj1 benchmark

rreusser commented:
I've made the changes and renamed the python benchmark.

kgryte commented:
Awesome. Thanks!

kgryte committed:
Merge branch 'develop' into besselj1

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/benchmark/r/benchmark.R:
```diff
@@ -0,0 +1,93 @@
+	#' @examples
+	#' print_summary( 3, 3 );
+	print_summary <- function( total, passing ) {
+		cat( "#\n" );
```
Looks like we still have indentation issues in this file.

rreusser replied:
I think I got it. I had to set `:noet` (no expand tab) and mess with tabstop in vim…

kgryte replied:
Oh nice. At some point (no rush), if you want to write a Vim guide (https://github.com/stdlib-js/stdlib/tree/develop/docs/editors) related to the project (linters and whatnot), that would be fab. :)

kgryte commented on lib/node_modules/@stdlib/math/base/special/besselj1/README.md:
```diff
@@ -0,0 +1,80 @@
+<!-- </equation> -->
+
+</section>
+
```
Missing closing `<!-- /.intro -->` comment.

rreusser replied:
Fixed.

kgryte reviewed (changes requested):
Two minor nits, but otherwise good on my end.

rreusser committed:
Add missing intro close in besselj1 readme

rreusser committed:
Fix tabs in besselj1 R benchmark

kgryte commented:
Once Circle CI finishes, should be ready for squash and merge! :)

rreusser commented on lib/node_modules/@stdlib/math/base/special/besselj0/README.md:
```diff
@@ -12,6 +12,8 @@ The [Bessel function of the first kind][bessel-first-kind] of order zero is defi
 
 </section>
 
+<!-- /.intro -->
```
Whoopsie. Well this one needs it too.

kgryte replied:
Yeah, went ahead and added it in the PR when I had to resolve the merge conflicts. :)
